plugin/file: zones files are opened but never closed inside fileParse

Reported by [Static Major from Orijtech Inc](https://github.com/marketplace/actions/staticmajor-analyzer), we can see this 
```shell
/coredns/plugin/file/setup.go:114:6: leaking resource created on line 91
/coredns/plugin/file/setup.go:118:6: leaking resource created on line 91
/coredns/plugin/file/setup.go:126:5: leaking resource created on line 91
```

which originates from https://github.com/coredns/coredns/blob/2f24b512d3999e9687ffbfaf8f8226c64185dca1/plugin/file/setup.go#L91 for which there is a missing reader.Close() which can be achieved by putting that code in an extracted closure/function and adding a `defer reader.Close()`

Kindly FYI-ing @willpoint @jhusdero @elias-orijtech @kirbyquerby

--- a/plugin/file/setup.go
+++ b/plugin/file/setup.go
@@ -92,19 +92,27 @@ func fileParse(c *caddy.Controller) (Zones, error) {
 		if err != nil {
 			openErr = err
 		}
-		defer reader.Close()
 
-		for i := range origins {
-			z[origins[i]] = NewZone(origins[i], fileName)
-			if openErr == nil {
-				reader.Seek(0, 0)
-				zone, err := Parse(reader, origins[i], fileName, 0)
-				if err != nil {
-					return Zones{}, err
+		err = func() error {
+			defer reader.Close()
+
+			for i := range origins {
+				z[origins[i]] = NewZone(origins[i], fileName)
+				if openErr == nil {
+					reader.Seek(0, 0)
+					zone, err := Parse(reader, origins[i], fileName, 0)
+					if err != nil {
+						return err
+					}
+					z[origins[i]] = zone
 				}
-				z[origins[i]] = zone
+				names = append(names, origins[i])
 			}
-			names = append(names, origins[i])
+			return nil
+		}()
+
+		if err != nil {
+			return Zones{}, err
 		}
 
 		for c.NextBlock() {

--- a/plugin/file/setup.go
+++ b/plugin/file/setup.go
@@ -92,6 +92,7 @@ func fileParse(c *caddy.Controller) (Zones, error) {
 		if err != nil {
 			openErr = err
 		}
+		defer reader.Close()
 
 		for i := range origins {
 			z[origins[i]] = NewZone(origins[i], fileName)

